From cbb7add59ddd0c67026942cbe6dd55fc094a35c4 Mon Sep 17 00:00:00 2001
From: Simon Gaynor <simon.gaynor@ezurio.com>
Date: Fri, 20 Sep 2024 09:37:23 -0700
Subject: [PATCH 1/2] hwmon: (ina2xx) add support for ina232

ina232 is similar to the ina226 except for bus_voltage_lsb
which is 1600mV and power_lsb_factor is 32

Signed-off-by: Simon Gaynor <simon.gaynor@ezurio.com>
---
 Documentation/hwmon/ina2xx.rst | 14 ++++++++++++--
 drivers/hwmon/ina2xx.c         | 19 ++++++++++++++++++-
 2 files changed, 30 insertions(+), 3 deletions(-)

Upstream-Status: Inappropriate No time for upstreaming

diff --git a/Documentation/hwmon/ina2xx.rst b/Documentation/hwmon/ina2xx.rst
index a3860aa..462e46e 100644
--- a/Documentation/hwmon/ina2xx.rst
+++ b/Documentation/hwmon/ina2xx.rst
@@ -53,6 +53,16 @@ Supported chips:
 
 	       https://www.ti.com/
 
+  * Texas Instruments INA232
+
+    Prefix: 'ina232'
+
+    Addresses: I2C 0x40 - 0x4f
+
+    Datasheet: Publicly available at the Texas Instruments website
+
+	       https://www.ti.com/
+
   * Texas Instruments INA260
 
     Prefix: 'ina260'
@@ -89,7 +99,7 @@ interface. The INA220 monitors both shunt drop and supply voltage.
 The INA226 is a current shunt and power monitor with an I2C interface.
 The INA226 monitors both a shunt voltage drop and bus supply voltage.
 
-INA230 and INA231 are high or low side current shunt and power monitors
+INA230, IN231 and INA232 are high or low side current shunt and power monitors
 with an I2C interface. The chips monitor both a shunt voltage drop and
 bus supply voltage.
 
@@ -106,7 +116,7 @@ compile-time or via the shunt_resistor attribute in sysfs at run-time. Please
 refer to the Documentation/devicetree/bindings/hwmon/ti,ina2xx.yaml for bindings
 if the device tree is used.
 
-Additionally ina226 supports update_interval attribute as described in
+Additionally ina226 and ina232 supports update_interval attribute as described in
 Documentation/hwmon/sysfs-interface.rst. Internally the interval is the sum of
 bus and shunt voltage conversion times multiplied by the averaging rate. We
 don't touch the conversion times and only modify the number of averages. The
diff --git a/drivers/hwmon/ina2xx.c b/drivers/hwmon/ina2xx.c
index bc3c1f7..3a160c6 100644
--- a/drivers/hwmon/ina2xx.c
+++ b/drivers/hwmon/ina2xx.c
@@ -18,6 +18,10 @@
  * Bi-directional Current/Power Monitor with I2C Interface
  * Datasheet: https://www.ti.com/product/ina230
  *
+ * INA232:
+ * Bi-directional Current/Power Monitor with I2C Interface
+ * Datasheet: https://www.ti.com/product/ina232
+ *
  * Copyright (C) 2012 Lothar Felten <lothar.felten@gmail.com>
  * Thanks to Jan Volkering
  */
@@ -135,7 +139,7 @@ static const struct regmap_config ina2xx_regmap_config = {
 	.writeable_reg = ina2xx_writeable_reg,
 };
 
-enum ina2xx_ids { ina219, ina226, ina260, sy24655 };
+enum ina2xx_ids { ina219, ina226, ina232, ina260, sy24655 };
 
 struct ina2xx_config {
 	u16 config_default;
@@ -184,6 +188,14 @@ static const struct ina2xx_config ina2xx_config[] = {
 		.has_ishunt = false,
 		.has_power_average = false,
 	},
+	[ina232] = {
+		.config_default = INA226_CONFIG_DEFAULT,
+		.calibration_value = 2048,
+		.shunt_div = 400,
+		.bus_voltage_shift = 0,
+		.bus_voltage_lsb = 1600,
+		.power_lsb_factor = 32,
+	},
 	[ina260] = {
 		.config_default = INA260_CONFIG_DEFAULT,
 		.shunt_div = 400,
@@ -990,6 +1002,7 @@ static const struct i2c_device_id ina2xx_id[] = {
 	{ "ina226", ina226 },
 	{ "ina230", ina226 },
 	{ "ina231", ina226 },
+	{ "ina232", ina232 },
 	{ "ina260", ina260 },
 	{ "sy24655", sy24655 },
 	{ }
@@ -1021,6 +1034,10 @@ static const struct of_device_id __maybe_unused ina2xx_of_match[] = {
 		.compatible = "ti,ina231",
 		.data = (void *)ina226
 	},
+	{
+		.compatible = "ti,ina232",
+		.data = (void *)ina232
+	},
 	{
 		.compatible = "ti,ina260",
 		.data = (void *)ina260
